src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshot.cs

```diff
@@ -15,32 +15,84 @@ namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
     // at once. 
     internal class DefaultProjectSnapshot : ProjectSnapshot
     {
-        public DefaultProjectSnapshot(Project underlyingProject)
+        public DefaultProjectSnapshot(string filePath, VersionStamp? version = null)
         {
-            if (underlyingProject == null)
+            if (filePath == null)
             {
-                throw new ArgumentNullException(nameof(underlyingProject));
+                throw new ArgumentNullException(nameof(filePath));
             }
 
-            UnderlyingProject = underlyingProject;
+            FilePath = filePath;
+            Version = version ?? VersionStamp.Default;
         }
 
-        private DefaultProjectSnapshot(Project underlyingProject, DefaultProjectSnapshot other)
+        public DefaultProjectSnapshot(HostProject hostProject, VersionStamp? version = null)
         {
-            if (underlyingProject == null)
+            if (hostProject == null)
             {
-                throw new ArgumentNullException(nameof(underlyingProject));
+                throw new ArgumentNullException(nameof(hostProject));
+            }
+
+            HostProject = hostProject;
+            
+            FilePath = hostProject.FilePath;
+            Version = version ?? VersionStamp.Default;
+        }
+
+        public DefaultProjectSnapshot(Project workspaceProject, VersionStamp? version = null)
+        {
+            if (workspaceProject == null)
+            {
+                throw new ArgumentNullException(nameof(workspaceProject));
+            }
+
+            WorkspaceProject = workspaceProject;
+            FilePath = workspaceProject.FilePath;
```

Comment:
Can we ever get into a state were the `FilePath` on `workspaceProject` is `null`? Asking because I'm trying to determine if it'd be valuable to solidify our system to understand snapshots that have a `null` file path (no host project or workspace project).

Reply:
You can deffo have projects in the workspace that don't have a path, but we don't support them.